url28dfd00382_byang223_TurnBAPI_tgz-pJ8-units_upgrades_UpgradeTestJ8/src/input/TBInputEngine.java

```diff
@@ -0,0 +1,240 @@
+	/**
+	 * Abstract method to be implemented for a specific game
+	 * to set default controls
+	 * @param map
+	 */
+	public abstract void createDefaultKeyBindingsMap(HashMap<String, Integer> map);
+	
+	/**
+	 * Abstract method determined by specific game
+	 * for deciding on the name of the Key Binding
+	 * file to save to
+	 * @return
+	 */
+	public abstract String getFileName();
+	
+	/**
+	 * Length of time that determines how
+	 * often to check if a file changed
+	 * @return
+	 */
+	public abstract int getFileCheckTime();
+
+	/**
+	 * Create FileWatcher to listen for file changes
+	 * Load key bindings depending on existence of file
+	 * NOTE: If file exists, must exist in correct format
+	 * @param g
+	 */
+	protected void loadEvents(TBGameMode g) {
+		File file=new File(getFileName());
+		TimerTask task = new FileWatcher(file, g);
+		Timer timer = new Timer();
+	    timer.schedule(task , new Date(), getFileCheckTime());
+	    if (!file.exists()) {
+	    	loadDefaultKeyBindings(g, file);
+	    }else{
+	    	loadKeyBindingsFromFile(g, file);
+	    }
+	}
+
+	/**
+	 * Reads from key bindings file and instantiates every class
+	 * with appropriate key
+	 * Assumes file is in correct format (does no error checking)
+	 * @param g
+	 * @param file
+	 * Reference: http://stackoverflow.com/questions/3671649/java-newinstance-of-class-thas-has-no-default-constructor
+	 */
+	protected void loadKeyBindingsFromFile(TBGameMode g, File file) {
+		getEventList().clear();
+		try{
+			FileInputStream fstream = new FileInputStream(file);
+			DataInputStream in = new DataInputStream(fstream);
+			BufferedReader br = new BufferedReader(new InputStreamReader(in));
+			String line;
+			while ((line = br.readLine()) != null)   {
+				String[] parts = line.split(" ");
+				Class<?> event = Class.forName(parts[0]);
+				Constructor<UserEvents> ctor = (Constructor<UserEvents>) event.getConstructor(TBGameMode.class, int.class);
+				getEventList().add(ctor.newInstance(g, Integer.parseInt(parts[1])));
+			}
+			in.close();
+		}catch (Exception e){
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * If no file exists, instantiates events based on 
+	 * user-defined database of key bindings.
+	 * Writes these key bindings to file.
+	 * @param g
+	 * @param file
+	 * Reference: http://stackoverflow.com/questions/3671649/java-newinstance-of-class-thas-has-no-default-constructor
+	 */
+	private void loadDefaultKeyBindings(TBGameMode g, File file) {
+		HashMap<String, Integer> keyBindingMap = new HashMap<String, Integer>();
+		createDefaultKeyBindingsMap(keyBindingMap);
+		for (String key : keyBindingMap.keySet()) {
+			Class<?> event;
+			try {
+				event = Class.forName(key);
+				Constructor<UserEvents> ctor = (Constructor<UserEvents>) event.getConstructor(TBGameMode.class, int.class);
+				getEventList().add(ctor.newInstance(g, keyBindingMap.get(key)));
+			} catch (Exception e) {
+				e.printStackTrace();
+			} 
+		}
+		
+		try{
+			FileWriter fstream = new FileWriter(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fstream*. The resource is referenced at lines 143, 145. The resource is closed at line 147. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *fstream* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).